### PR TITLE
Add Batch JSON-RPC support (rpc client & server)

### LINKED ISSUE #329: [rpc] Add support for JSON-RPC 2.0 batch requests
The current RPC server only supports JSON-RPC 1.0 requests.  The JSON-RPC 2.0 spec added several things, one of which is the ability to batch several requests into one.  We need to add support for these batched requests.

I have some code which does this for the normal path if anyone gets to this before I get back around to it and wants a starting point, however there are various considerations that need to be made for supporting batching in the websockets path.  In particular, the websockets path supports the notion of some long-running requests, currently only rescan, executing asynchronously so the faster commands can continue to be executed while the long-running operation is performed asynchronously.  Thus, the interaction with batching must be carefully considered when one of the requests in the batch is one that is executed asynchronously.

### PR DESCRIPTION
**Pull request by jakesyl**
This pr aims to address #329 which has been open for a while now by adding bulk request support. This is base on from [decred#841](https://github.com/decred/dcrd/pull/841) by @davecgh

### CODE REVIEW

**Comment by jakesyl** [comment]
I think we need some sort of bulk handler we can tack other methods on too here.

**Comment by torkelrogstad** [comment]
@jakesyl Do you want review on this now, or should I hold off until it's less of a WIP?

**Comment by jakesyl** [comment]
@torkelrogstad, let's hold off for now. I'll have this done by EoD. Feel free to check it out in the meantime. I think I know how I want to solve the bulk querying in post mode now.

**Comment by jakesyl** [comment]
Actually @torkelrogstad, I'd love to get your input on the best way to queue these bulk requests. Here's how it's done in python: 

image

I'm playing with how to do it in btcd

**Comment by jakesyl** [comment]
I'm thinking of adding a batch method onto client that returns a modified client. Here's the [aforementioned implementation](https://github.com/jgarzik/python-bitcoinrpc/blob/master/bitcoinrpc/authproxy.py#L149)

**Comment by jakesyl** [comment]
@torkelrogstad I'm sure we need some work here, but this is what I've got so far. Please review and let me know what you think

**Comment by torkelrogstad** [comment]
I've started looking it over, great stuff! Review is still pending, I'll wait until I've finished the entire PR before publishing. Hopefully I'll find some time later today.

**Comment by jakesyl** [comment]
Any updates on this?

**Comment by jakesyl** [comment]
Hey, any idea if any changes are required here?

File: rpcserver.go [hunk 1]
```diff
@@ -3947,80 +4009,165 @@ func (s *rpcServer) jsonRPCRead(w http.ResponseWriter, r *http.Request, isAdmin
...
-					Code:    btcjson.ErrRPCInvalidParams.Code,
-					Message: "limited user not authorized for this method",
+		}
+
+		if err == nil {
+			// Response with an empty batch error if the batch size is zero
+			if len(batchedRequests) == 0 {
+				jsonErr := &btcjson.RPCError{
+					Code:    btcjson.ErrRPCInvalidRequest.Code,
+					Message: fmt.Sprint("Invalid request: empty batch"),
```
Comment (torkelrogstad) [inline rpcserver.go]: can use string literal directly, no need for `fmt.Sprint`

File: rpcwebsocket.go [hunk 2]
```diff
@@ -1323,153 +1323,435 @@ out:
...
 			}
 
-			reply, err := createMarshalledReply(cmd.id, nil, cmd.err)
-			if err != nil {
-				rpcsLog.Errorf("Failed to marshal parse failure "+
-					"reply: %v", err)
+			if req.Method == "" || req.Params == nil {
+				jsonErr := &btcjson.RPCError{
+					Code:    btcjson.ErrRPCInvalidRequest.Code,
+					Message: fmt.Sprintf("Invalid request: malformed"),
```
Comment (torkelrogstad) [inline rpcwebsocket.go]: can use string literal directly

File: rpcwebsocket.go [hunk 3]
```diff
@@ -1323,153 +1323,435 @@ out:
...
+			if err == nil {
+				// Response with an empty batch error if the batch size is zero
+				if len(batchedRequests) == 0 {
+					if !c.authenticated {
+						break out
+					}
+
+					jsonErr := &btcjson.RPCError{
+						Code:    btcjson.ErrRPCInvalidRequest.Code,
+						Message: fmt.Sprint("Invalid request: empty batch"),
```
Comment (torkelrogstad) [inline rpcwebsocket.go]: can use string literal directly

File: rpcwebsocket.go [hunk 4]
```diff
@@ -1323,153 +1323,435 @@ out:
...
+							if reply != nil {
+								results = append(results, reply)
+							}
+							continue
+						}
+
+						if req.Method == "" || req.Params == nil {
+							jsonErr := &btcjson.RPCError{
+								Code:    btcjson.ErrRPCInvalidRequest.Code,
+								Message: fmt.Sprintf("Invalid request: malformed"),
```
Comment (torkelrogstad) [inline rpcwebsocket.go]: can use string literal directly

File: rpcclient/chain.go [hunk 5]
```diff
@@ -982,3 +981,31 @@ func (c *Client) GetCFilterHeader(blockHash *chainhash.Hash,
...
+	Id     uint64      `json:"id"`
+}
+
+type BulkResult = map[uint64]IndividualBulkResult
+
+// Receive waits for the response promised by the future and returns the hash of
+// the best block in the longest block chain.
+func (r FutureGetBulkResult) Receive() (BulkResult, error) {
+	m := make(BulkResult)
+	res, err := receiveFuture(r)
```
Comment (torkelrogstad) [inline rpcclient/chain.go]:
```go
res, err := receiveFuture(r)
if err != nil {
	return nil, err
}
```

**Comment by jakesyl** [comment]
Didn't get a notification about this, I'll go through these changes tonight!

File: rpcclient/infrastructure.go [hunk 6]
```diff
@@ -1441,3 +1474,64 @@ func (c *Client) BackendVersion() (BackendVersion, error) {
 
 	return *c.backendVersion, nil
 }
+
+// make batch requests
+func (c *Client) Batch() *Client {
+	c.batch = true //copy the client with changed batch setting
```
[the remark this answers is not in the document]
Comment (jakesyl) [reply rpcclient/infrastructure.go]: Yes, what are your thoughts here?

**Comment by jakesyl** [comment]
@torkelrogstad all set

File: btcjson/jsonrpc.go [hunk 7]
```diff
@@ -73,15 +73,65 @@ type Request struct {
...
+	var data map[string]interface{}
+	err := json.Unmarshal(b, &data)
+	if err != nil {
+		return err
+	}
+
+	request.ID = data["id"]
+	methodValue, hasMethod := data["method"]
+	if hasMethod {
+		request.Method = methodValue.(string)
```
[the remark this answers is not in the document]
Comment (jakesyl) [reply btcjson/jsonrpc.go]: Interesting point. I don't see an alternative though

File: rpcclient/examples/bitcoincorehttpbulk/README.md [hunk 8]
```diff
@@ -0,0 +1,32 @@
+Bitcoin Core Batch HTTP POST Example
+==============================
+
+This example shows how to use the rpcclient package to connect to a Bitcoin
+Core RPC server using HTTP POST mode with TLS disabled and uses the batch json rpc mode
+
+## Running the Example
```
[the remark this answers is not in the document]
Comment (jakesyl) [reply rpcclient/examples/bitcoincorehttpbulk/README.md]: I don't think so, this isn't done anywhere else in the project. Seems out of scope

**Comment by jakesyl** [comment]
Any luck getting this merged

**Comment by jakesyl** [comment]
Hi, any update here?

**Comment by jakesyl** [comment]
Looks like CI is failing, will check

**Comment by jakesyl** [comment]
For whatever reason, this is working locally and not on ci (running `./goclean.sh`)

**Comment by jakesyl** [comment]
image

**Comment by torkelrogstad** [comment]
@jakesyl are you sure IDEA/Goland is running the the `rpctest` tag enabled?

**Comment by jakesyl** [comment]
@torkelrogstad, yes. But on top of that it is also passing in `./goclean.sh`

**Comment by pavel-main** [comment]
@jakesyl I've occasionally discovered that it's related to `GOPATH` caching - whenever I remove the line, build succeeds https://github.com/btcsuite/btcd/pull/1536/commits/fe003e236663978c5c792e0ed077d5e83b4db6be

**Comment by jakesyl** [comment]
@pavel-main thanks! Trying this now

**Comment by jakesyl** [comment]
Thanks @pavel-main, looks like that fixde it!

**Comment by jakesyl** [comment]
Any update here? Looks like prereq #1537 has been fixed

**Comment by jakesyl** [comment]
@jcvernaleo 

- low priority

**Comment by jakesyl** [comment]
Going to need to squash

File: btcjson/btcdextcmds_test.go [hunk 9]
```diff
@@ -193,7 +193,7 @@ func TestBtcdExtCmds(t *testing.T) {
 	for i, test := range tests {
 		// Marshal the command as created by the new static command
 		// creation function.
-		marshalled, err := btcjson.MarshalCmd(testID, test.staticCmd())
+		marshalled, err := btcjson.MarshalCmd("1.0", testID, test.staticCmd())
```
Comment (jcvernaleo) [inline btcjson/btcdextcmds_test.go]: Why are you changing from a variable to a hardcoded number here (and all the places below?

Comment (jakesyl) [reply btcjson/btcdextcmds_test.go]: fair point, I'll have to refactor. I've held off on squashing till these revisions are complete, but can do this earlier fi it helps your review

Comment (jakesyl) [reply btcjson/btcdextcmds_test.go]: Appreciate the feedback

Comment (jcvernaleo) [reply btcjson/btcdextcmds_test.go]: Generally I agree about waiting to squash until revisions are done but this needs the rebase before reviewing it for real so you might as well squash at the same time (I hate making people do extra steps if I don't have to).

Comment (jakesyl) [reply btcjson/btcdextcmds_test.go]: No problem

**Comment by jakesyl** [comment]
@jcvernaleo I merged this by accident, went ahead and reverted instantly.

**Comment by jakesyl** [comment]
can I force push to undo?

**Comment by jakesyl** [comment]
@jcvernaleo merged the wrong branch here - I thought this was on my local. I undid immediately after (#1549), but can't push the history here because the branch is protected :(, please advise

**Comment by jakesyl** [comment]
image
@davecgh tagging you here as well

**Comment by jakesyl** [comment]
btw: 
image

**Comment by onyb** [comment]
@jakesyl Are you still actively working on this?

**Comment by jakesyl** [comment]
Hi @onyb yes - I've been very busy the past month but have a few PRs queued here including this one that I need to work on. I'm still active in the btcd irc

File: rpcclient/infrastructure.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (onyb) [reply rpcclient/infrastructure.go:1480]: It doesn't seem to be making a copy of the struct. As a consequence, if you perform a synchronous RPC call using the original client, the program will hang.


Here's a small snippet from the example here: `rpcclient/examples/bitcoincorehttpbulk/main.go`
```go
batchClient := client.Batch()  // mutates client

// batch mode requests are fine
blockCount := batchClient.GetBlockCountAsync()
difficulty := batchClient.GetDifficultyAsync()

// sends all queued batch requests
batchClient.Send()

fmt.Println(blockCount.Receive())
fmt.Println(difficulty.Receive())

fmt.Println(client == batchClient)   // CAUTION: expected false, got true
fmt.Println(client.GetDifficulty())  // CAUTION: program hangs
```

**Comment by jakesyl** [comment]
Recreated in #1583

File: btcjson/jsonrpc.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (jakesyl) [reply btcjson/jsonrpc.go:89]: @torkelrogstad I've gone ahead and addressed this on #1583 

image

**Pull request merged**